### ISSUE TITLE
View assignment error for 1d view

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#3766]
This code:

```c++
#include <Kokkos_Core.hpp>
#include <Kokkos_DualView.hpp>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv); {

  Kokkos::DualView<int[3], Kokkos::Cuda> v("v");

  Kokkos::View<int[3], Kokkos::DefaultHostExecutionSpace> a = v.h_view;

  } Kokkos::finalize();

}

```
gives the error about incompatible view assignment due to memory layout:

```c++
/home/aznb/mycodes/testKokkos/build_cuda_debug/_deps/kokkos-src/core/src/Kokkos_View.hpp(1452): error: static assertion failed with "Incompatible View copy construction"
          detected during instantiation of "Kokkos::View<DataType, Properties...>::View(const Kokkos::View<RT, RP...> &, std::enable_if<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, void>::type *) [with DataType=int [3], Properties=<Kokkos::DefaultHostExecutionSpace>, RT=int [3], RP=<Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>>]" 
/home/aznb/mycodes/testKokkos/testViewAssign.cpp(18): here

/home/aznb/mycodes/testKokkos/build_cuda_debug/_deps/kokkos-src/core/src/impl/Kokkos_ViewMapping.hpp(3381): error: static assertion failed with "View assignment must have compatible layout or have rank <= 1"
          detected during:
            instantiation of "void Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::assign(Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::DstType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::SrcType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::TrackType &) [with DstTraits=Kokkos::ViewTraits<int [3], Kokkos::DefaultHostExecutionSpace>, SrcTraits=Kokkos::ViewTraits<int [3], Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>>]" 
/home/aznb/mycodes/testKokkos/build_cuda_debug/_deps/kokkos-src/core/src/Kokkos_View.hpp(1454): here
            instantiation of "Kokkos::View<DataType, Properties...>::View(const Kokkos::View<RT, RP...> &, std::enable_if<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, void>::type *) [with DataType=int [3], Properties=<Kokkos::DefaultHostExecutionSpace>, RT=int [3], RP=<Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>>]" 
/home/aznb/mycodes/testKokkos/testViewAssign.cpp(18): here

/home/aznb/mycodes/testKokkos/build_cuda_debug/_deps/kokkos-src/core/src/impl/Kokkos_ViewMapping.hpp(1755): error: static assertion failed with "ViewOffset LayoutRight and LayoutLeft are only compatible when rank <= 1"
          detected during:
            instantiation of "Kokkos::Impl::ViewOffset<Dimension, Kokkos::LayoutRight, std::enable_if<<expression>, void>::type>::ViewOffset(const Kokkos::Impl::ViewOffset<DimRHS, Kokkos::LayoutLeft, void> &) [with Dimension=Kokkos::Impl::ViewDimension<3UL>, DimRHS=Kokkos::Impl::ViewDimension<3UL>]" 
(3427): here
            instantiation of "void Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::assign(Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::DstType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::SrcType &, const Kokkos::Impl::ViewMapping<DstTraits, SrcTraits, std::enable_if<<expression>, void>::type>::TrackType &) [with DstTraits=Kokkos::ViewTraits<int [3], Kokkos::DefaultHostExecutionSpace>, SrcTraits=Kokkos::ViewTraits<int [3], Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>>]" 
/home/aznb/mycodes/testKokkos/build_cuda_debug/_deps/kokkos-src/core/src/Kokkos_View.hpp(1454): here
            instantiation of "Kokkos::View<DataType, Properties...>::View(const Kokkos::View<RT, RP...> &, std::enable_if<Kokkos::Impl::ViewMapping<Kokkos::View<DataType, Properties...>::traits, Kokkos::View<RT, RP...>::traits, Kokkos::ViewTraits<DataType, Properties...>::specialize>::is_assignable_data_type, void>::type *) [with DataType=int [3], Properties=<Kokkos::DefaultHostExecutionSpace>, RT=int [3], RP=<Kokkos::Cuda::array_layout, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::HostSpace::memory_space>>]" 
/home/aznb/mycodes/testKokkos/testViewAssign.cpp(18): here

```
I thought layout left vs layout right shouldn't matter for 1D view. Also, changing from `int[3]` to `int*` avoids the error.

**Comment by crtrott** [comment]
oh interesting. Apparently we are missing some case there.

**Comment by nliber** [comment]
In [Kokkos_ViewMapping.hpp](https://github.com/kokkos/kokkos/blob/develop/core/src/impl/Kokkos_ViewMapping.hpp)

We have a static assert at [Line 1751](https://github.com/kokkos/kokkos/blob/ca723a6a7a88bd35561bdf896dce025fcd017f13/core/src/impl/Kokkos_ViewMapping.hpp#L1751):

```
  template <class DimRHS>
  KOKKOS_INLINE_FUNCTION constexpr ViewOffset(
      const ViewOffset<DimRHS, Kokkos::LayoutLeft, void>& rhs)
      : m_dim(rhs.m_dim.N0, 0, 0, 0, 0, 0, 0, 0) {
    static_assert((DimRHS::rank == 0 && dimension_type::rank == 0) ||
                      (DimRHS::rank == 1 && dimension_type::rank == 1 &&
                       dimension_type::rank_dynamic == 1),
                  "ViewOffset LayoutRight and LayoutLeft are only compatible "
                  "when rank <= 1");
  }

```

It is unclear why `dimension_type::rank_dynamic == 1` is needed.

Similarly, [Line 3344](https://github.com/kokkos/kokkos/blob/ca723a6a7a88bd35561bdf896dce025fcd017f13/core/src/impl/Kokkos_ViewMapping.hpp#L3344) also checks this:

```
  enum {
    is_assignable_layout =
        std::is_same<typename DstTraits::array_layout,
                     typename SrcTraits::array_layout>::value ||
        std::is_same<typename DstTraits::array_layout,
                     Kokkos::LayoutStride>::value ||
        (DstTraits::dimension::rank == 0) ||
        (DstTraits::dimension::rank == 1 &&
         DstTraits::dimension::rank_dynamic == 1)
  };

```

If the `rank_dynamic` check from both these places is removed the code appears to work.

**Comment by Char-Aznable** [comment]
@nliber it seems reasonable to remove the `rank_dynamic` check. Wish I could help with a PR but I just don't feel comfortable with Kokkos' internal

**Comment by PhilMiller** [comment]
This is fixed on `develop`, sometime since 3.7 was released.
